src/execution/__tests__/semantic-nullability-test.ts

```diff
@@ -0,0 +1,219 @@
+    const data = {
+      a: () => 'Apple',
+      b: () => {
+        throw new Error('Something went wrong');
+      },
+      c: () => 'Cookie',
+    };
+
+    const document = parse(`
+        query {
+          b
+        }
+      `);
+
+    const executable = document.definitions?.values().next()
+      .value as ExecutableDefinitionNode;
+    const selectionSet = executable.selectionSet.selections
+      .values()
+      .next().value;
+
+    const result = await execute({
+      schema: new GraphQLSchema({ query: DataType }),
+      document,
+      rootValue: data,
+    });
+
+    expect(result).to.deep.equal({
+      data: {
+        b: null,
+      },
+      errors: [
+        new GraphQLError('Something went wrong', {
+          nodes: selectionSet,
+          path: ['b'],
+        }),
+      ],
+    });
+  });
+
+  it('SemanticNonNull halts null propagation', async () => {
+    const deepData = {
+      f: () => null,
+    };
+
+    const data = {
+      a: () => 'Apple',
+      b: () => null,
+      c: () => 'Cookie',
+      d: () => deepData,
+    };
+
+    const document = parse(`
+        query {
+          d {
+            f
+          }
+        }
+      `);
+
+    const result = await execute({
+      schema: new GraphQLSchema({ query: DataType }),
+      document,
+      rootValue: data,
+    });
+
+    const executable = document.definitions?.values().next()
+      .value as ExecutableDefinitionNode;
+    const dSelectionSet = executable.selectionSet.selections.values().next()
+      .value as FieldNode;
+    const fSelectionSet = dSelectionSet.selectionSet?.selections
+      .values()
+      .next().value;
+
+    expect(result).to.deep.equal({
+      data: {
+        d: null,
+      },
+      errors: [
+        new GraphQLError(
+          'Cannot return null for non-nullable field DeepDataType.f.',
+          {
+            nodes: fSelectionSet,
+            path: ['d', 'f'],
+          },
+        ),
+      ],
+    });
+  });
+
+  it('SemanticNullable allows null values', async () => {
+    const data = {
+      a: () => null,
+      b: () => null,
+      c: () => 'Cookie',
+    };
+
+    const document = parse(`
+        query {
+          a
+        }
+      `);
+
+    const result = await execute({
+      schema: new GraphQLSchema({ query: DataType }),
+      document,
+      rootValue: data,
+    });
+
+    expect(result).to.deep.equal({
+      data: {
+        a: null,
+      },
+    });
+  });
+
+  it('SemanticNullable allows non-null values', async () => {
+    const data = {
+      a: () => 'Apple',
+      b: () => null,
+      c: () => 'Cookie',
+    };
+
+    const document = parse(`
+        query {
+          a
+        }
+      `);
+
+    const result = await execute({
+      schema: new GraphQLSchema({ query: DataType }),
+      document,
+      rootValue: data,
+    });
+
+    expect(result).to.deep.equal({
+      data: {
+        a: 'Apple',
+      },
+    });
+  });
```

Comment:
I don't think these test cases are necessary anymore now that `SemanticNullable` has been removed.